### ISSUE TITLE
BadDigest: The Content-MD5 you specified did not match what we received.

### ISSUE THREAD

**Issue by johnf** [electron-userland/electron-builder#1772]
* **Version**: 19.11.0

* **Target**:  Windows, Mac and Linux

Since upgrading to 19.11.0 I'm getting BadDigest issues when uploading to S3. Rolling back to 19.8.0 and the problem goes away. I'm assuming it's related to b57dc8a9e47a79eaaefb083cf176e4ff365837a0

Please let me know if there is anything I can do to help debug.

```
hubber-setup-0.0.5.exe to GitHubBadDigest: The Content-MD5 you specified did not match what we received.
    at Request.extractError (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/services/s3.js:577:35)
    at Request.callListeners (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/sequential_executor.js:105:20)
    at Request.emit (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/request.js:682:14)
    at Request.transition (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/request.js:684:12)
    at Request.callListeners (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/sequential_executor.js:115:18)
    at Request.emit (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/request.js:682:14)
    at Request.transition (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/request.js:684:12)
    at Request.callListeners (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/sequential_executor.js:115:18)
    at callNextListener (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/sequential_executor.js:95:12)
    at IncomingMessage.onEnd (/Users/travis/build/echohubio/hubber/node_modules/aws-sdk/lib/event_listeners.js:269:13)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
From previous event:
    at Uploader.putObject (/Users/travis/build/echohubio/hubber/node_modules/electron-publisher-s3/src/uploader.ts:96:5)
    at Timeout.tryRun [as _onTimeout] (/Users/travis/build/echohubio/hubber/node_modules/electron-publisher-s3/src/uploader.ts:198:9)
    at ontimeout (timers.js:365:14)
    at tryOnTimeout (timers.js:237:5)
    at Timer.listOnTimeout (timers.js:207:5)
error Command failed with exit code 255.
```

**Comment by johnf** [comment]
So I'm seeing this on travis but only on Mac OS X not on Linux

**Comment by develar** [comment]
Solution — we should migrate to S3 managed upload.

**Comment by develar** [comment]
I doubt that it is our bug — it seems it is really md5 mismatch.

**Comment by johnf** [comment]
Yeah I'm trying to debug it now

**Comment by develar** [comment]
BTW, please ensure that you use latest electron-publisher-s3 19.11.0

**Comment by johnf** [comment]
Zeroing in. The MD5 code is fine. The problem seems to be that the ```mac-latest.{yml,json}``` is being written two multiple times and each time the date is slightly different so the MD5Sum changes.

My config is

``` yaml
  "build": {
    "mac": {
      "category": "public.app-category.productivity"
    },
    "publish": [
      {
        "provider": "generic",
        "url": "https://packages.staging.echohub.io/${os}"
      },
      {
        "provider": "s3",
        "bucket": "echohub-packages-staging",
        "path": "${os}"
      },
      {
        "provider": "github"
      }
    ]
  },
```

In this case mac is builfin DMG and ZIP. If I limit to one of those the problem goes away but that could just be a timing thing.

Some debug output
```
Signing app (identity:  Developer ID Application: )
Building macOS zip
Publishing to null
Publishing to S3 (bucket: echohub-packages-staging)
Publishing to Github (owner: echohubio, project: hubber, version: 0.0.5)
UPDATING: mac.json 2017-07-01T00:20:51.197Z
Publishing to null
UPDATING: mac.yml 2017-07-01T00:20:51.550Z
[                    ] 0% 0.0s | hubber-0.0.5-mac.zip to GitHub
Publishing to null
UPDATING: mac.json 2017-07-01T00:20:52.108Z
UPDATING: mac.yml 2017-07-01T00:20:52.114Z
# MD5 Calc starts here
# FIle is updates in the middle
UPDATING: mac.json 2017-07-01T00:20:52.131Z
# Cal is finished
MD5 Calc: /Users/travis/build/echohubio/hubber/dist/latest-mac.json (G+J/DHBEEgr1Sf+Nagf6Pw==)


[====================] 100% 0.0s | latest-mac.json to S3
UPDATING: mac.yml 2017-07-01T00:20:52.149Z

MD5 Calc: /Users/travis/build/echohubio/hubber/dist/latest-mac.yml (RkrrpGl8RmwbIOGKjK05IA==)
[====================] 100% 0.0s | hubber-0.0.5-mac.zip to GitHub


[====================] 100% 0.0s | latest-mac.json to GitHub
[====================] 100% 0.0s | latest-mac.yml to S3
[====================] 100% 0.0s | latest-mac.yml to GitHub
[====================] 100% 0.0s | hubber-0.0.5-mac.zip to S3
BadDigest: The Content-MD5 you specified did not match what we received.

```

You can see above that the MD5 function starts and just before the hash is calculated the file is written to again

**Comment by johnf** [comment]
OK so the problem I'm hitting is here
https://github.com/electron-userland/electron-builder/blob/master/packages/electron-builder/src/publish/PublishManager.ts#L232-L257

It is because I'm publishing to multiple places. So for every location the latest files are regenerated.

**Comment by johnf** [comment]
A possibly naive fix exists  #1775

**Comment by joshuahiggins** [comment]
I am still experiencing this. I'm running the latest:

```
"electron-builder": "19.15.1",
"electron-publisher-s3": "19.15.0"
```

I'm attempting to deploy to both S3 and Github and receiving the same `BadDigest` error mentioned above.

**Comment by joshuahiggins** [comment]
I'm still experiencing this on `electron-builder@19.15.1` `electron-publisher-s3@19.15.0`. I've attempted to deploy to 2 separate S3 folders as well as to S3 and Github. Each time I get a `BadDigest` error.

**Comment by develar** [comment]
@joshuahiggins Will be fixed in the 19.5.4.

**Comment by tylerlong** [comment]
I can reproduce the issue with electron-builder 19.45.0. And I confirm it goes away with electron-builder 19.45.4.

Great work!
